modules/rostests/apitests/ntdll/NtQueryInformationProcess.c

```diff
@@ -323,6 +325,130 @@ Test_ProcessWx86Information(void)
     trace("VdmPower = %lu\n", VdmPower);
 }
 
+static
+void
+Test_ProcQueryAlignmentProbe(void)
+{
+    NTSTATUS Status, SpecialStatus = STATUS_SUCCESS;
+    ULONG InfoClass;
+
+    /* Iterate over the process info classes and begin the tests */
+    for (InfoClass = 0; InfoClass < _countof(PsProcessInfoClass); InfoClass++)
+    {
+        /*
+         * Switch over specific information classes that return specific (special) status
+         * values when tested against Windows Server 2003. These respective status values
+         * indicate that the information class table data are correct and Windows Server 2003
+         * goes further into the code execution until NtQueryInformationProcess() fails
+         * for whatever reason on these classes mentioned within the switch body.
+        */
+        switch (InfoClass)
+        {
+            case ProcessWorkingSetWatch:
+            {
+                SpecialStatus = STATUS_UNSUCCESSFUL;
+                break;
+            }
+
+            case ProcessHandleTracing:
+            {
+                SpecialStatus = STATUS_INVALID_PARAMETER;
+                break;
+            }
+
+            /*
+             * This class returns an arbitrary size pointed by InformationLength
+             * which equates to the image filename of the process. Such status
+             * is returned in an invalid address query (STATUS_ACCESS_VIOLATION)
+             * where the function expects STATUS_INFO_LENGTH_MISMATCH instead.
+            */
+            case ProcessImageFileName:
+            {
+                SpecialStatus = STATUS_INFO_LENGTH_MISMATCH;
+                break;
+            }
+        }
+
+        /* The buffer is misaligned */
+        Status = NtQueryInformationProcess(NtCurrentProcess(),
+                                           InfoClass,
+                                           (PVOID)(ULONG_PTR)(PsProcessInfoClass[InfoClass].AlignmentQUERY / 2),
+                                           PsProcessInfoClass[InfoClass].RequiredSizeQUERY,
+                                           NULL);
+
+        /*
+         * In Windows Server 2003 this function doesn't perform
+         * class validity check for query when simulating a
+         * misaligned buffer case scenario. However there are
+         * classes that are implemented in Vista+ where
+         * Windows Server 2003 outright fails on them
+         * with STATUS_INVALID_INFO_CLASS.
+        */
+        if (!(PsProcessInfoClass[InfoClass].Flags & ICIF_QUERY) && !(PsProcessInfoClass[InfoClass].Flags & ICIF_SET))
```

Comment:
So what's interesting here is that our `PsProcessInfoClass` table will of course have `AlignmentQUERY=0` for these classes.
Therefore, since `0` is an aligned address, we're not really testing here whether an alignment check is performed.
It would be interesting to see how the function behaves for those classes when given an actually unaligned pointer, e.g.
```
if (PsProcessInfoClass[InfoClass].Flags & ICIF_QUERY)
{
    Pointer = (PVOID)(ULONG_PTR)(PsProcessInfoClass[InfoClass].AlignmentQUERY / 2); // as above
}
else
{
    Pointer = (PVOID)(ULONG_PTR)1;
}
```